qiskit/qasm3/exporter.py

```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+                reserving a dummy variable name that does not actually have a Qiskit object backing
+                it.
+            allow_rename: whether to allow the name to be mutated to escape it and/or make it safe
+                to define (avoiding keywords, subject to shadowing rules, etc).
+            global_: force this declaration to be in the global scope.
+        """
+        scope_index = 0 if global_ else -1
+        # We still need to do this escaping and shadow checking if `global_`, because we don't want
+        # a previous variable declared in the currently active scope to shadow the global.  This
+        # kind of logic would be cleaner if we made the naming choices later, after AST generation
+        # (e.g. by using only indices as the identifiers until we're ready to output the program).
+        name = self.escaped_declarable_name(name, allow_rename=allow_rename, unique=global_)
+        identifier = ast.Identifier(name)
+        self.variables[scope_index][name] = variable
+        if variable is not None:
+            self.objects[scope_index][variable] = identifier
+        return identifier
+
+    def set_object_ident(self, ident: ast.Identifier, variable: object):
+        """Set the identifier used to refer to a given object for this scope.
+
+        This overwrites any previously set identifier, such as during the original registration.
+
+        This is generally only useful for tracking "sub" objects, like bits out of a register, which
+        will have an `SubscriptedIdentifier` as their identifier."""
+        self.objects[-1][variable] = ident
+
+    def get_variable(self, variable: object) -> ast.Identifier:
+        """Lookup a non-gate variable in the symbol table."""
+        for scope in reversed(self.objects):
+            if (out := scope.get(variable)) is not None:
+                return out
+        raise KeyError(f"'{variable}' is not defined in the current context")
+
+    def register_gate_without_definition(self, name: str, gate: Gate | None) -> ast.Identifier:
+        """Register a gate that does not require an OQ3 definition.
+
+        If the ``gate`` is given, it will be used to validate that a call to it is compatible (such
+        as a known gate from an included file).  If it is not given, it is treated as a user-defined
+        "basis gate" that assumes that all calling signatures are valid and that all gates of this
+        name are exactly compatible, which is somewhat dangerous."""
+        # Validate the name is usable.
+        name = self.escaped_declarable_name(name, allow_rename=False)
+        ident = ast.Identifier(name)
+        if gate is None:
+            self.gates[name] = GateDefinition(None, None)
+        else:
+            source = _gate_definition_source(gate)
+            self.gates[name] = GateDefinition(source, None)
+            if (standard_gate := getattr(source, "_standard_gate")) is not None:
```

Comment:
Why is it `getattr` here, but `._standard_gate` in `_gate_definition_source`?

Reply:
I'm not entirely sure.  Everything's required to be `Gate` (or `None`, but that's already handled) throughout these paths, so the attribute always exists.  Removed unnecessary `getattr` in b51386fac2.